### ISSUE TITLE
CPU multi threading

### ISSUE THREAD

**Issue by jafdy** [endless-sky/endless-sky#4045]
Related: #387

Since turrets were introduced there have been significantly more AI calculations required, potentially resulting in larger gains (to maximum ships before frame time slows) than there would have been about this time 3 years ago. 
Currently the game runs on 2 threads, engine and graphical, this results in most modern computers being under utilised with only the lowest end new chips having 2 core (4 threads). While this won't benefit older computers nearly as much, it will allow the game's CPU processing ability to continue to grow with the newer many core CPUs that are likely coming in the future.

I would regard this as at best low priority but I feel that it should be considered again.

**Comment by MCOfficer** [comment]
disclaimer: i have no experience with c++, so take anything i say with a grain of salt (or two).

it seems that each frame, the game iterates over all ships and adjusts their aim, fires their guns etc.

in other programming languages (which is where my experience comes from), iterating over a collection and doing the same thing (call `::Step`) is a very simple way of multithreading. however, things get potentially thread-unsave when these objects interact with each other (e.g. thead A's ship fires at a ship whose actions are currently being calculated by thread B).

simple actions that only require read interactions (accessing a ship's coordinates rather than *changing* it's shield strength) could be done in multiple threads - however that also means that ship A (which was at the front of the queue) can only react to ship Z's actions in the next frame.

that, however, also has the consequence that the code needs to be restructured, since currently all potentially thread-safe methods are called by `::Step`, which also calls thread-unsafe methods.

bottom line, this is a large, potentially very risky and work-intensive task - multithreading a tasks that were meant to be single-threaded always is.

**Comment by tehhowch** [comment]
No one had ever said "we won't multithread Endless Sky" -- it just hasn't been done yet. 

Arguably, MZ has worked towards this in the past year with various refactoring steps. It is worth remembering that multithreading is not free, and has some associated overhead with dispatching work and collecting results.

The two most obvious candidates are the ship loop in AI::Step, and the projectile loop 
(collision detection). I know as fact that there is an existing order of operations dependence in the AI calculation (a parent ship has always decided what it will do before its escorts decide, and they refer to its decisions) and such order is no longer guaranteed once we partition that work.

**Comment by Zitchas** [comment]
In some respects, having ships get a mechanically enforced delay of not being able to react to other ships until the next frame is kind of a good thing. Realistically, none of us can react to what other ships are doing simultaneously to them starting to do it. Having the AI get a 1/60th of a second delay (or whatever a frame ends up being) would be a small improvement to making the  other ships feel like they were flown by people instead of automated systems with perfect timing. Admittedly, such a small delay is almost certainly too small for any of us to notice, so it would be pretty much irrelevant. It does illustrate the point, however, that introducing a certain amount of delay there would be a more-than-acceptable cost if the system as a whole could then handle a lot more complexity and/or entities.

**Comment by endless-sky** [comment]
In general, the cost of splitting a task among multiple threads is high enough that it only makes sense if the task is relatively time-consuming. All the computations, combined, fit into 1/60th of a second, and the AI is an even smaller fraction of that. So there's a very real risk that thread-switching costs would swamp out any performance gains (or that some OSes would get faster, and others would get slower).

I agree with @tehhowch that the collision detection would be relatively easy to multithread. As for the AI, the costliest part is the turret auto-aiming and auto-firing. That could easily be reduced by only having each turret make the auto-firing decision a few times per second (e.g. on each frame only 1/6 or 1/10th of all the guns check whether they should fire). Auto-tracking turrets could also only make the choice of which target they're tracking on that same schedule.

**Comment by tux2603** [comment]
I am currently without a major project. Has there been any progress on this issue recently?

**Comment by Zitchas** [comment]
As far as I've seen, there hasn't been any mention of this. I would recommend stopping by the Discord to talk with tehhowch or the other coders.

**Comment by Xyncgas** [comment]
SIMD might be better than async or multi-thread at doing these calculations

**Comment by tibetiroka** [comment]
SIMD is good at some tasks, but isn't universally available (x86-64 only guarantees SSE2, whereas we require SSE3 for efficiently vectorizing Point operations). It also isn't great at performing all of the necessary operations, and we have to be careful with it due to some locks in the dictionary code.

**Comment by Zitchas** [comment]
> AI, the costliest part is the turret auto-aiming and auto-firing. That could easily be reduced by only having each turret make the auto-firing decision a few times per second (e.g. on each frame only 1/6 or 1/10th of all the guns check whether they should fire). Auto-tracking turrets could also only make the choice of which target they're tracking on that same schedule.

Could this be as simple as just having a simple `If RANDOM>25 then do CanFireCheck` to bypass that chunk of code 75% of the time?

Just a thought: It would be interesting if this tied into the faction inaccuracy. Supposedly that is supposed to represent how trained a particular faction is, and thus how accurate in a firefight. Makes sense that the gunner in a Navy turret has better situational awareness and capable of making those checks more often than, say, the civilian merchant ship crew member who really is just trained to run a civilian ship but gets stuck in the turret now and then.

That might make the pseudo code something like `if RANDOM > (25+inaccuracy) then do IsBestTargetCheck.` So less professional factions have higher inaccuracy, and thus lower chance of doing the "Is this the best target for me to shoot at?" calculations. The same check could be done for "Can I shoot now?" although honestly, if this was implemented, then the turret should *keep* shooting until it does a check and gets a "No, nothing to shoot at". This would make higher inaccuracy factions prone to shooting after the target has left weapons range. 

Just looking at human fleets. Merchants have 20-50 confusion; pirates have about 20. No other fleets in the human fleet file have any. So with the above values, Navy ships recalculate 75% of the time. Pirates and militia types about 55% of frames, merchants 50-25% of frames. 

Heh, if we wanted to make drones more common, it would be interesting to put confusion onto ships, and specify things like the cheap human drones having confusion 50 (that applies only to the "should I recheck targeting" questions) so they effectively only recheck for targets about once per second. (of course, advanced drones probably wouldn't have this)

((As an aside, I'd also like to see faction inaccuracy used to cause a ship to start shooting before the target is within range, as well as sometimes lagging and not starting to shoot until it is well within range, but that's not for CPU threading, that's just changing behavior)
